### PR TITLE
Feature/field

### PR DESCRIPTION
**Pull request by nikitinalex**
Made field constructed only be string and IOC as singleton. Also rewrote field test for new logic. For IOC-container added methods for adding and removing transformation rules. Also Added tests for it.

### CODE REVIEW

File: src/main/java/info/smart_tools/smartactors/utils/ioc/IOC.java [hunk 1]
```diff
@@ -17,7 +22,28 @@
...
-    public IOC() {
+    //Singleton instance
+    private static volatile IOC INSTANCE = null;
+
+    /**
+     * Gives instance if class exists or otherwise creates it.
+     * Realized with double checked locking
+     * @return instance of IOC container
+     */
+    public static IOC getInstance() {
```
Comment (Menliat) [inline src/main/java/info/smart_tools/smartactors/utils/ioc/IOC.java]: I thinks it's simpler to use instance holder idiom for thread-safe singleton implementation: https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom

File: src/main/java/info/smart_tools/smartactors/core/receivers/invocation/InvokersFactory.java [hunk 2]
```diff
@@ -19,11 +19,9 @@
  * TODO: support for multiple arguments for method (using argument names as field names)
  */
 public class InvokersFactory {
-    private IOC iocContainer;
     private MessageWrapperGenerator generator;
 
-    public InvokersFactory(IOC iocContainer, MessageWrapperGenerator generator) {
-        this.iocContainer = iocContainer;
+    public InvokersFactory(MessageWrapperGenerator generator) {
```
Comment (Menliat) [inline src/main/java/info/smart_tools/smartactors/core/receivers/invocation/InvokersFactory.java]: I'm not sure, but maybe it's easier to leave this file unchanged in order to reduce the number of hidden implicit dependencies introduced by making singleton from IOC container.
The main goal is to be able to use Fields without pacing IOC container/field instances to constructor of every class. Making singleton from IOC is a hard and non-clear choise, so I think we should reduce the impact of it as much as possible.

File: src/main/java/info/smart_tools/smartactors/core/Field.java [hunk 3]
```diff
@@ -6,23 +6,21 @@
 public class Field<T> {
 
     private FieldName name;
-    private IOC ioc;
 
-    public Field(FieldName name, IOC ioc) {
```
Comment (Menliat) [inline src/main/java/info/smart_tools/smartactors/core/Field.java]: Wouldn't it be simpler to test if we make another constructor by FieldName which calls the previous one augmented with IOC.getInstance()?

**Comment by Menliat** [comment]
Can we try to reduce usage of IOC container singleton instance in system tests? It makes tests dependent on each other and may cause strange tests failures when we try to run them in a different order (especially when we have tests for handling of errors by IOC container with some fake registrations).

Generally speaking, I think it's reasonable to create separate IOC container instance for fields which will contain only known pure mappings from one type to another (like array -> list, primitive types conversion, etc) and another IOC container for initialization of the whole system. 
To my mind, type conversion tasks (done by Fields) differ from the work done by IOC container, since the container instantiates *parts of the system* and Fields map *input to the system* from one type to another. For example, it's quite hard to imagine (and work with) fields like Field<Mailbox> or Field<Router> or Field<Actor>, because it's not a kind of objects which can be carried by IMessage and it's not explicit how these object were created (it matters when an object has a lot of different-purpose implementations with side-effects, like MessageReceiver).
Anyway, I guess we can postpone this separation (if it sound rational right now).

**Comment by nikitinalex** [comment]
At the moment we have more important and urgent problems that we should solve. So That's a good point about separating IOC-containers, but we'll make it later. As for other remarks: tests and constructors returned to the start position and IOC singleton made with Demand Holder idiom.

**Comment by Menliat** [comment]
Sounds good! I guess, we can merge it.

**Pull request merged**